add technote on Apollo Router resource requests/limits (#264)
apollographql/docs: src/content/technotes/TN0015-router-resource-management.mdx

---

> title: Apollo Router resource management

Reviewer: ```suggestion
title: Managing Apollo Router resources in Kubernetes
```

Reviewer: Might help w/ search and also zero in on content scope

id: TN0017
tags: [router, best practices, kubernetes]
---

Determining the correct resource requests and limits for your application pods in a Kubernetes system is not an exact science. Your specific needs depend on many factors, including:

> - The cardinality of unique operation shapes.
> - The latency of underlying subgraphs and data sources.
> - The size of responses.
> - The complexity of query plans.

Reviewer: ```suggestion
- The cardinality of unique operation shapes
- The latency of underlying subgraphs and data sources
- The size of responses
- The complexity of query plans
```

Reviewer: (We don't end-punctuate list items unless they're complete sentences)

Our general recommendation for Kubernetes is to start with these requests and limits:

```yaml
resources:
  requests:
    memory: '1gb'
    cpu: '1000m'
  limits:
    memory: '2gb'
    # no CPU limit to avoid throttling
```

> When using [Horizontal Pod Autoscaling](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), we recommend targeting 90% utilization:

Reviewer: ```suggestion
When using [Horizontal pod autoscaling](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/), we recommend targeting 90% utilization:
```

```yaml
metrics:
  - type: Resource
    resource:
      name: cpu
      target:
        type: Utilization
        averageUtilization: 90
  - type: Resource
    resource:
      name: memory
      target:
        type: Utilization
        averageUtilization: 90
```

> Apollo Router starts up quickly, but when using [Managed Federation](https://www.apollographql.com/docs/router/managed-federation/overview/), the Router has to fetch the Supergraph schema from Apollo Uplink before it can start serving traffic. We recommend measuring your Router startup time and lowering the `averageUtilization` if your startup time is longer due to Uplink latency or the size of your Supergraph.

Reviewer: ```suggestion
The Apollo Router starts up quickly, but when using [Managed Federation](/router/managed-federation/overview/), the router has to fetch the supergraph schema from Apollo Uplink before it can start serving traffic. We recommend measuring your router's startup time and lowering the `averageUtilization` if your startup time is longer due to Uplink latency or the size of your supergraph schema.
```
